src/devices/cpu/sh/sh7014_dmac.cpp

```diff
@@ -286,25 +304,25 @@ TIMER_CALLBACK_MEMBER( sh7014_dmac_channel_device::dma_timer_callback )
 		m_dar -= m_active_dma_unit_size;
 
 	if (m_request_source_type == RS_TYPE_INTERNAL)
-		m_notify_dma_source_cb(m_selected_resource);
+		m_dmac->notify_dma_source(m_selected_resource);
 
 	switch (m_active_dma_unit_size) {
 		case 1:
-			m_cpu->m_program->write_byte(
+			m_dmac->m_cpu->m_program->write_byte(
 				m_dar,
-				m_cpu->m_program->read_byte(m_sar)
+				m_dmac->m_cpu->m_program->read_byte(m_sar)
 			);
 			break;
 		case 2:
-			m_cpu->m_program->write_word(
+			m_dmac->m_cpu->m_program->write_word(
 				m_dar,
-				m_cpu->m_program->read_word(m_sar)
+				m_dmac->m_cpu->m_program->read_word(m_sar)
 			);
 			break;
 		case 4:
-			m_cpu->m_program->write_dword(
+			m_dmac->m_cpu->m_program->write_dword(
 				m_dar,
-				m_cpu->m_program->read_dword(m_sar)
+				m_dmac->m_cpu->m_program->read_dword(m_sar)
```

Comment:
@cuavas While reworking this code I wanted to simplify the channels and stop passing down the CPU device and INTC device into each DMA channel, but hit a little bit of a snag here. Do you think it'd be better to just go back to passing in the CPU device and INTC device instead of accessing the private variables inside the parent DMA controller device? Or maybe it'd be better to just unlink them entirely (no longer friended)?

Reply:
Got some sleep and looked at this again and it still feels bad relying on private internals like this so I went back to just passing in m_cpu and m_intc to the channels since it feels like the less bad of the two.

I'm still not satisfied with holding m_cpu and m_intc inside sh7014_dmac_device because they're only ever used to pass into the channels though, but the only other thing I could think of is exposing a set_cpu and set_intc from the channel and using a lookup to set the CPU/INTC for the channels from sh2_sh7014_device::device_add_mconfig. 

In the end I'm just not sure how to best structure this so feedback would still be appreciated here if there's a recommended way.